### ISSUE TITLE
pre-allocation of disk space

### ISSUE THREAD

**Issue by playgithub** [yt-dlp/yt-dlp#9685]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Take youtube for example, `yt-dlp --list-formats`, it shows the size for a specific format, so it's possible to pre-allocate disk space.
BTW, though the size might not be accurate, a workaround is possible.

### Provide verbose output that clearly demonstrates the problem

- [ ] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [ ] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

**Comment by pukkandan** [comment]
Can you explain your usecase?

**Comment by playgithub** [comment]
> Can you explain your usecase?

The download speed for me is very slow, allocate disk space KB by KB is not so efficient, and while other applications are allocating disk space at the same time, the physical allocation of file on disk may not be so continuous. If the pre-allocation doesn't write disk, it'll get some benifits by pre-allocation.

BTW many downloader supports pre-allocated disk space.

**Comment by pukkandan** [comment]
> The download speed for me is very slow

I doubt this will be solved by pre-allocation. Any reason to believe otherwise? Any example we can test? (with verbose logs)

**Comment by playgithub** [comment]
> > The download speed for me is very slow
> 
> I doubt this will be solved by pre-allocation. Any reason to believe otherwise? Any example we can test? (with verbose logs)

I don't mean pre-allocation will speed up downloading, I mean slow downloading speed has chances to make the downloading file on the disk not so continuous. e.g. when download several big files simultaneously, it may like this: file A writes some bytes on disk, then file B, then file C, ..., as a result the files on the disk might interweave each other on the disk.

**Comment by Grub4K** [comment]
@playgithub: So your concern is fragmentation? I don't think that will matters much for everyday usage at all, even if using HDDs. Please give an example where fragmentation would be detrimental.
As a workaround you may use `aria2c` with allows to preallocate disk space afaik.

---
@pukkandan: Note that downloaders usually provide the preallocation option to not run out of space while downloading (has happened to me before as well). This I think is a valid concern, however this would be hard to implement for the reason that no bigger disk writes should occur after preallocation (state file would be fine). I think at least the fragmented downloader will have problems flushing directly to file offset.

**Comment by playgithub** [comment]
> @playgithub: So your concern is fragmentation? I don't think that will matters much for everyday usage at all, even if using HDDs. Please give an example where fragmentation would be detrimental.

Yes, fragmentation.
With pre-allocation, there's only good and no bad.
However if it is complex to implement (especially for cross-platform), it might be better not to implement it.

> As a workaround you may use `aria2c` with allows to preallocate disk space afaik.

That's fine.

**Comment by pukkandan** [comment]
I was not implying pre-allocation is useless; but want to understand OP's use-case and whether it'll actually be solved by pre-allocation. This issue [looks like an XY Problem](https://xyproblem.info) to me since there may be various symptoms that a user unfamiliar with yt-dlp's internals assume will be solved by pre-allocation, but actually won't.

It's also not trivial to implement for us, with the various interactions with resume options, fragmentation, postprocessing and needing to be cross-platform. So I want to ensure the benefits are actually worth the effort.

Correct me if I am wrong in any of these - these are my initial concerns, not tested conclusions:

First off, we cannot pre-allocate for fragmented files since the final file size is not known in advance. There are low level strategies to deal with such situation, but they are much more complex that a pre-allocation.

Most post-processing - especially the ones that need a lot of writes - are done by ffmpeg. I don't believe ffmpeg has a pre-allocation option.

So, what's left is direct downloads with no postprocessing; for which fragmentation is not a significant issue unless performing a lot of downloads simultaneously (not recommended for network/disk contention reasons anyway) - a very narrow use-case imo.

The easiest workaround for fragmentation concerns is just to download to a different disk using `-P temp:`. If you can use a ramdisk like @​voidpointertonull, all the better.

**Comment by playgithub** [comment]
I think buffering to write large chunks is good enough.
Does `yt-dlp` support it already by option `--http-chunk-size`?

**Comment by sergeevabc** [comment]
+1 for pre-allocation switch!

**Comment by flawrist** [comment]
@veganomy 

I use SSD only for short-term work, and I trust downloading and storage to HDD, which is why I am so delighted when a developer adds "pre-allocation of disk space" option (for example, in [Tixati](https://tixati.com/) torrent client), and I am sad when the app has no such feature.

**Comment by veganomy** [comment]
Nobody is promoting or complaining here about magnetic drives or solid state drives. You're free to choose your own way of storing files.

Your usecase was to avoid slow download speeds, as you described before. Even though you've pre-allocated the storage memory into linear links, the download speeds of Hard Drives can still be rate limited via different pathways such as network speed, YouTube servers, your geolocation, background applications, etc. Lowest specifications are always the biggest bottleneck when it comes to computer hardware. You upgrade them, you raise the lower limits.

**Comment by ddawson** [comment]
@veganomy

> Your usecase was to avoid slow download speeds,

That's not what playgithub said. If you [read carefully](#issuecomment-2053991594), you can see the goal is to *mitigate the problems* that can come with a slow download, specifically that there are more opportunities for other things to get written to disk in the meantime, which can lead to more fragmentation. Pre-allocation should help with that, though whether the problem is severe enough to be worth the trouble mitigating is another matter.

**Comment by veganomy** [comment]
Downloads are already linearly allocated, when you download http streams because they're just one continuous part file. The problem only exists on dash/hls streams, as you get fragmented pieces, and then a merge function.

After completing the merger, the fragments are automatically discarded, which means those inode memory address locations are reallocatable again.

There's another issue with preallocation. None of the stream metadata actually provides accurate filesizes. It's just a mere approximation through bitrate and other forms of metadata. Using this data to preallocate local storage, might lead to disk failure errors.

**Comment by playgithub** [comment]
> After completing the merger, the fragments are automatically discarded, which means those inode memory address locations are reallocatable again.

Yes, btw it's nice for an option to download the intermediate contents to memory only, but more work, it needs to integrate ffmpeg as a lib.

**Comment by veganomy** [comment]
If you're on Linux, guess what. You're in luck. Just mount a tmpfs, write a script to download there first & copy back to your downloads directory.

> tmpfs is a ramdisk storage, which is 10x faster than your fastest NAND storage read/write speeds.

**Comment by playgithub** [comment]
> If you're on Linux, guess what. You're in luck. Just mount a tmpfs, write a script to download there first & copy back to your downloads directory.
> 
> > tmpfs is a ramdisk storage, which is 10x faster than your fastest NAND storage read/write speeds.

Convenient for linux users. However downloading to memory directly (without memdisk) is more user friendly and crossplatform.

**Comment by tcely** [comment]
FYI, the [SABR branch](https://github.com/coletdjnz/yt-dlp-dev/tree/feat/youtube/sabr) has a nice backend system that should help with this.

When this pull request is merged, it would be worth looking at way to download to memory:

- #13515 

The segment / sequence architecture should solve the problem of not knowing the accurate file size before downloading in a convenient way. An assembly algorithm could be created to pre-allocate based on the total size.
